### PR TITLE
CA1070: Do not declare event fields as virtual

### LINKED ISSUE #3410: Do not mark field events as virtual
Following [MS guidelines](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/events/how-to-raise-base-class-events-in-derived-classes):

> Do not declare virtual events in a base class and override them in a derived class. The C# compiler does not handle these correctly and it is unpredictable whether a subscriber to the derived event will actually be subscribing to the base class event.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #3410

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtual.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterSymbolAction(context =>
+            {
+                var eventSymbol = (IEventSymbol)context.Symbol;
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtual.cs]: Shall I implement the `api_surface` option?

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtual.cs:39]: Normally design rules only execute on public API surface by default. Seems reasonable to do the same for this rule and support the `api_surface` option for customization.

Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtual.cs:39]: Will do

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/xlf/MicrosoftCodeQualityAnalyzersResources.cs.xlf [hunk 2]
```diff
@@ -97,6 +97,21 @@
         <target state="translated">Nezachytávejte obecné typy výjimek</target>
         <note />
       </trans-unit>
+      <trans-unit id="DoNotDeclareEventFieldsAsVirtualDescription">
+        <source>Do not declare virtual events in a base class and override them in a derived class. The C# compiler does not handle these correctly and it is unpredictable whether a subscriber to the derived event will actually be subscribing to the base class event.</source>
+        <target state="new">Do not declare virtual events in a base class and override them in a derived class. The C# compiler does not handle these correctly and it is unpredictable whether a subscriber to the derived event will actually be subscribing to the base class event.</target>
+        <note />
+      </trans-unit>
+      <trans-unit id="DoNotDeclareEventFieldsAsVirtualMessage">
+        <source>Do not declare event fields as virtual.</source>
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/xlf/MicrosoftCodeQualityAnalyzersResources.cs.xlf:106]: Message should likely include the name of the event symbol.
```suggestion
        <source>Event '{0}' should not be declared virtual</source>
```

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtualTests.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        [Fact]
+        public async Task EventPropertyVirtual_NoDiagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+public class C
+{
+    public event EventHandler ThresholdReached
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtualTests.cs:32]: Did you mean to declare this as virtual? Is that a valid syntax?

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtualTests.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        [Fact]
+        public async Task EventFieldVirtual_Diagnostic()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+public class C
+{
+    public virtual event EventHandler ThresholdReached;
+}",
+                VerifyCS.Diagnostic().WithLocation(5, 39));
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtualTests.cs:22]: Why not markup syntax?

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtualTests.cs (hunk 3 above)
Comment (Evangelink) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtualTests.cs:32]: Yes and yes :)

File: src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtualTests.cs (hunk 4 above)
Comment (Evangelink) [reply src/NetAnalyzers/UnitTests/Microsoft.CodeQuality.Analyzers/QualityGuidelines/DoNotDeclareEventFieldsAsVirtualTests.cs:22]: I might be wrong in my understanding but the markup syntax is not testing the various properties like the `VerifyCS..Diagnostic()` does. Is it right @sharwell?

File: src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx [hunk 5]
```diff
@@ -1381,6 +1381,14 @@
   </data>
   <data name="DoNotNameEnumValuesReservedTitle" xml:space="preserve">
     <value>Do not name enum values 'Reserved'</value>
+  <data name="DoNotDeclareEventFieldsAsVirtualDescription" xml:space="preserve">
+    <value>Do not declare virtual events in a base class and override them in a derived class. The C# compiler does not handle these correctly and it is unpredictable whether a subscriber to the derived event will actually be subscribing to the base class event.</value>
+  </data>
+  <data name="DoNotDeclareEventFieldsAsVirtualMessage" xml:space="preserve">
+    <value>Do not declare event fields as virtual.</value>
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/MicrosoftCodeQualityAnalyzersResources.resx:1388]: Do you mean to update the message? I see the analyzer does pass in the event name as the argument to reported diagnostic now..

File: src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md [hunk 6]
```diff
@@ -7,6 +7,7 @@ CA1021 | Design | Disabled | AvoidOutParameters, [Documentation](https://docs.mi
 CA1045 | Design | Disabled | DoNotPassTypesByReference, [Documentation](https://docs.microsoft.com/visualstudio/code-quality/ca1045)
 CA1047 | Design | Info | DoNotDeclareProtectedMembersInSealedTypes, [Documentation](https://docs.microsoft.com/visualstudio/code-quality/ca1047)
 CA1069 | Design | Info | EnumShouldNotHaveDuplicatedValues, [Documentation](https://docs.microsoft.com/visualstudio/code-quality/ca1069)
+CA1070 | Design | Info | DoNotDeclareEventFieldsAsVirtual, [Documentation](https://docs.microsoft.com/visualstudio/code-quality/ca1070)
```
Comment (mavasani) [inline src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:10]: Please file a doc issue to add documentation for this rule.

Comment (Evangelink) [reply src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:10]: https://github.com/MicrosoftDocs/visualstudio-docs/issues/5151

**Pull request merged**